### PR TITLE
En 8009/gas schedule soft forks

### PR DESCRIPTION
**Pull request by sasurobert**
implemented gas schedule soft fork mechanism with the same style as epoch notifier.
every component registers to the gas schedule notifier and will get the complete new gas map when that is changed.

### CODE REVIEW

File: process/smartContract/builtInFunctions/esdtBurn.go [hunk 1]
```diff
@@ -45,6 +45,11 @@ func NewESDTBurnFunc(
 	return e, nil
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (e *esdtBurn) SetNewGasConfig(gasCost *process.GasCost) {
+	e.funcGasCost = gasCost.BuiltInCost.ESDTBurn
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/esdtBurn.go:50]: We need to protect `e.funcGasCost` with a mutex

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go [hunk 2]
```diff
@@ -22,6 +22,11 @@ func NewClaimDeveloperRewardsFunc(gasCost uint64) *claimDeveloperRewards {
 	return &claimDeveloperRewards{gasCost: gasCost}
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (c *claimDeveloperRewards) SetNewGasConfig(gasCost *process.GasCost) {
+	c.gasCost = gasCost.BuiltInCost.ClaimDeveloperRewards
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/claimDeveloperRewards.go:27]: We need to protect `e.gasCost` with a mutex

File: process/smartContract/builtInFunctions/changeOwnerAddress.go [hunk 3]
```diff
@@ -21,6 +21,11 @@ func NewChangeOwnerAddressFunc(gasCost uint64) *changeOwnerAddress {
 	return &changeOwnerAddress{gasCost: gasCost}
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (c *changeOwnerAddress) SetNewGasConfig(gasCost *process.GasCost) {
+	c.gasCost = gasCost.BuiltInCost.ChangeOwnerAddress
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/changeOwnerAddress.go:26]: We need to protect `e.gasCost` with a mutex

File: process/smartContract/builtInFunctions/esdtTransfer.go [hunk 4]
```diff
@@ -51,6 +51,11 @@ func NewESDTTransferFunc(
 	return e, nil
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (e *esdtTransfer) SetNewGasConfig(gasCost *process.GasCost) {
+	e.funcGasCost = gasCost.BuiltInCost.ESDTBurn
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/esdtTransfer.go:56]: We need to protect `e.funcGasCost` with a mutex

File: process/smartContract/builtInFunctions/factory.go [hunk 5]
```diff
@@ -14,120 +14,177 @@ var log = logger.GetOrCreate("process/smartContract/builtInFunctions")
...
 	MapDNSAddresses      map[string]struct{}
 	EnableUserNameChange bool
 	Marshalizer          marshal.Marshalizer
 	Accounts             state.AccountsAdapter
 }
 
-// CreateBuiltInFunctionContainer will create the list of built-in functions
-func CreateBuiltInFunctionContainer(args ArgsCreateBuiltInFunctionContainer) (process.BuiltInFunctionContainer, error) {
-	gasConfig, err := createGasConfig(args.GasMap)
+type builtInFuncFactory struct {
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/factory.go:24]: Maybe increase the code coverage for this?

File: process/smartContract/builtInFunctions/keyValueStorage.go [hunk 6]
```diff
@@ -32,6 +32,12 @@ func NewSaveKeyValueStorageFunc(
 	return s, nil
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (k *saveKeyValueStorage) SetNewGasConfig(gasCost *process.GasCost) {
+	k.funcGasCost = gasCost.BuiltInCost.SaveKeyValue
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/keyValueStorage.go:37]: We need to protect `k.funcGasCost` and `k.gasConfig` with a mutex

File: process/smartContract/builtInFunctions/saveUserName.go [hunk 7]
```diff
@@ -41,6 +41,11 @@ func NewSaveUserNameFunc(
 	return s, nil
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (s *saveUserName) SetNewGasConfig(gasCost *process.GasCost) {
+	s.gasCost = gasCost.BuiltInCost.SaveUserName
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/saveUserName.go:46]: We need to protect `s.gasCost` with a mutex

File: process/smartContract/process.go [hunk 8]
```diff
@@ -175,10 +175,22 @@ func NewSmartContractProcessor(args ArgsNewSmartContractProcessor) (*scProcessor
...
 }
 
+// GasScheduleChange sets the new gas schedule where it is needed
+func (sc *scProcessor) GasScheduleChange(gasSchedule map[string]map[string]uint64) {
+	apiCosts := gasSchedule[core.ElrondAPICost]
+	if apiCosts == nil {
+		return
+	}
+
+	sc.asyncCallStepCost = apiCosts[core.AsyncCallStepField]
```
Comment (iulianpascalau) [inline process/smartContract/process.go:190]: We need to protect `sc.asyncCallStepCost` and  `sc.asyncCallbackGasLock` with a mutex
Since we are providing the same map pointer this is prone to race conditions. Why not replacing `map[string]map[string]uint64` with and interface that has only some getters with some internal mutex that will protect the node even if it will be used deliberately wrong?

File: process/transaction/transactionCostEstimator.go [hunk 9]
```diff
@@ -44,6 +44,11 @@ func NewTransactionCostEstimator(
 	}, nil
 }
 
+// GasScheduleChange is called when gas schedule is changed, thus all contracts must be updated
+func (tce *transactionCostEstimator) GasScheduleChange(gasSchedule map[string]map[string]uint64) {
+	tce.compilePerByteCost, tce.storePerByteCost = getOperationCost(gasSchedule)
```
Comment (iulianpascalau) [inline process/transaction/transactionCostEstimator.go:49]: We need to protect these with a mutex

File: vm/process/systemVM.go [hunk 10]
```diff
@@ -129,6 +129,17 @@ func (s *systemVM) RunSmartContractCall(input *vmcommon.ContractCallInput) (*vmc
...
 }
 
+// GasScheduleChange sets the new gas schedule where it is needed
+func (s *systemVM) GasScheduleChange(gasSchedule map[string]map[string]uint64) {
+	apiCosts := gasSchedule[core.ElrondAPICost]
+	if apiCosts == nil {
+		return
+	}
+
+	s.asyncCallStepCost = apiCosts[core.AsyncCallStepField]
```
Comment (iulianpascalau) [inline vm/process/systemVM.go:139]: We need to protect these with a mutex

File: vm/systemSmartContracts/auction.go [hunk 11]
```diff
@@ -1435,6 +1435,11 @@ func (s *stakingAuctionSC) getBlsKeysStatus(args *vmcommon.ContractCallInput) vm
 	return vmcommon.Ok
 }
 
+// SetNewGasCosts is called whenever a gas cost was changed
+func (s *stakingAuctionSC) SetNewGasCosts(gasCost vm.GasCost) {
+	s.gasCost = gasCost
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/auction.go:1440]: We need to protect `s.gasCost` with a mutex

File: vm/systemSmartContracts/esdt.go [hunk 12]
```diff
@@ -795,6 +795,11 @@ func (e *esdt) EpochConfirmed(epoch uint32) {
 	log.Debug("esdt contract", "enabled", e.flagEnabled.IsSet())
 }
 
+// SetNewGasCosts is called whenever a gas cost was changed
+func (e *esdt) SetNewGasCosts(gasCost vm.GasCost) {
+	e.gasCost = gasCost
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:800]: We need to protect `s.gasCost` with a mutex

File: vm/systemSmartContracts/governance.go [hunk 13]
```diff
@@ -910,6 +910,11 @@ func (g *governanceContract) EpochConfirmed(epoch uint32) {
 	log.Debug("governance contract", "enabled", g.flagEnabled.IsSet())
 }
 
+// SetNewGasCosts is called whenever a gas cost was changed
+func (g *governanceContract) SetNewGasCosts(gasCost vm.GasCost) {
+	g.gasCost = gasCost
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/governance.go:915]: We need to protect `s.gasCost` with a mutex

File: vm/systemSmartContracts/staking.go [hunk 14]
```diff
@@ -1427,6 +1427,11 @@ func (r *stakingSC) EpochConfirmed(epoch uint32) {
 	log.Debug("stakingSC: stake/unstake/unbond", "enabled", r.flagStake.IsSet())
 }
 
+// SetNewGasCosts is called whenever a gas cost was changed
+func (r *stakingSC) SetNewGasCosts(gasCost vm.GasCost) {
+	r.gasCost = gasCost
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1432]: We need to protect `s.gasCost` with a mutex

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: cmd/node/main.go [hunk 15]
```diff
@@ -191,12 +191,11 @@ VERSION:
...
 		Value: "./config/p2p.toml",
 	}
-	// gasScheduleConfigurationFile defines a flag for the path to the toml file containing the gas costs used in SmartContract execution
-	gasScheduleConfigurationFile = cli.StringFlag{
-		Name: "gas-costs-config",
-		Usage: "The `" + filePathPlaceholder + "` for the gas costs configuration file. This TOML file contains " +
-			"gas costs used in SmartContract execution",
-		Value: "./config/gasSchedule.toml",
+	// gasScheduleConfigurationFolder defines a flag for the path to the folder containing the gas costs used in execution
+	gasScheduleConfigurationFolder = cli.StringFlag{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/main.go:195]: done.

File: core/forking/gasSchedule.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+type GasScheduleMap = map[string]map[string]uint64
+
+type gasScheduleNotifier struct {
+	mutNotifier       sync.RWMutex
+	configDir         string
+	gasScheduleConfig config.GasScheduleConfig
+	currentEpoch      uint32
+	lastGasSchedule   GasScheduleMap
+
+	handlers []core.GasScheduleSubscribeHandler
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/forking/gasSchedule.go:24]: done

File: core/forking/gasSchedule.go [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+		log.Error("could not load the new gas schedule")
+		return
+	}
+
+	log.Debug("gasScheduleNotifier.EpochConfirmed new gas schedule",
+		"new epoch", epoch,
+		"num handlers", len(g.handlers),
+	)
+
+	g.lastGasSchedule = newGasSchedule
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/forking/gasSchedule.go:111]: this change can be done only from epochnotifier change epoch -> which is called in the main node process. There are no concurrent calls here.

[the remark this answers is not in the document]
Comment (sasurobert) [reply core/forking/gasSchedule.go:111]: it is protected all the way right now.

File: core/forking/gasSchedule.go [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+// EpochConfirmed is called whenever a new epoch is confirmed
+func (g *gasScheduleNotifier) EpochConfirmed(epoch uint32) {
+	old := atomic.SwapUint32(&g.currentEpoch, epoch)
+	sameEpoch := old == epoch
+	if sameEpoch {
+		return
+	}
+
+	g.mutNotifier.RLock()
+	newVersion := g.getMatchingVersion(epoch)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/forking/gasSchedule.go:92]: done

File: integrationTests/vm/arwen/utils.go [hunk 19]
```diff
@@ -100,6 +100,7 @@ type testParticipant struct {
 	BalanceSnapshot *big.Int
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/vm/arwen/utils.go:100]: this is for tests only.

File: process/smartContract/builtInFunctions/esdtBurn.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtBurn.go:50]: no need. there is no possibility to call the from different go routines

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go (hunk 2 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:27]: no need. there is no possibility to call the from different go routines

File: process/smartContract/builtInFunctions/changeOwnerAddress.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/changeOwnerAddress.go:26]: no need. there is no possibility to call the from different go routines

File: process/smartContract/builtInFunctions/esdtTransfer.go (hunk 4 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtTransfer.go:56]: no need. there is no possibility to call the from different go routines

File: process/smartContract/builtInFunctions/factory.go (hunk 5 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/factory.go:24]: the factory is tested for most of the cases. it will error in case of bad arguments and it is tested.

File: process/smartContract/builtInFunctions/keyValueStorage.go (hunk 6 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/keyValueStorage.go:37]: no need. there is no possibility to call the from different go routines

File: process/smartContract/builtInFunctions/saveUserName.go (hunk 7 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/saveUserName.go:46]: no need. there is no possibility to call the from different go routines

File: process/smartContract/process.go (hunk 8 above)
Comment (sasurobert) [reply process/smartContract/process.go:190]: no need. there is no possibility to call the from different go routines

Comment (sasurobert) [reply process/smartContract/process.go:190]: because if there is an interface it needs to be added like everywhere, vmcommon, vm-util, arwen any change here or there makes it impossible to use.

Comment (sasurobert) [reply process/smartContract/process.go:190]: there are no mutexes inside smartContract/process.go at all.

File: vm/process/systemVM.go (hunk 10 above)
Comment (sasurobert) [reply vm/process/systemVM.go:139]: no need. there is no possibility to call the from different go routines

File: process/transaction/transactionCostEstimator.go (hunk 9 above)
Comment (sasurobert) [reply process/transaction/transactionCostEstimator.go:49]: yes. this will be protected as it is accessible through api.

File: vm/systemSmartContracts/auction.go (hunk 11 above)
Comment (sasurobert) [reply vm/systemSmartContracts/auction.go:1440]: no need. there is no possibility to call the from different go routines

File: vm/systemSmartContracts/esdt.go (hunk 12 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:800]: no need. there is no possibility to call the from different go routines

File: vm/systemSmartContracts/governance.go (hunk 13 above)
Comment (sasurobert) [reply vm/systemSmartContracts/governance.go:915]: no need. there is no possibility to call the from different go routines

File: vm/systemSmartContracts/staking.go (hunk 14 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1432]: no need. there is no possibility to call the from different go routines

File: core/forking/gasSchedule.go [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+	g := &gasScheduleNotifier{
+		gasScheduleConfig: gasScheduleConfig,
+		handlers:          make([]core.GasScheduleSubscribeHandler, 0),
+	}
+
+	sort.Slice(g.gasScheduleConfig.GasScheduleByEpochs, func(i, j int) bool {
+		return g.gasScheduleConfig.GasScheduleByEpochs[i].StartEpoch < g.gasScheduleConfig.GasScheduleByEpochs[j].StartEpoch
+	})
+	var err error
+	g.lastGasSchedule, err = core.LoadGasScheduleConfig(filepath.Join(configDir, gasScheduleConfig.GasScheduleByEpochs[0].FileName))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/forking/gasSchedule.go:42]: done

File: process/smartContract/builtInFunctions/esdtBurn.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtBurn.go:50]: it is protected all the way right now.

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go (hunk 2 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:27]: it is protected all the way right now.

File: process/smartContract/builtInFunctions/changeOwnerAddress.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/changeOwnerAddress.go:26]: protected all the way

File: process/smartContract/builtInFunctions/esdtTransfer.go (hunk 4 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtTransfer.go:56]: protected all the way

File: process/smartContract/builtInFunctions/keyValueStorage.go (hunk 6 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/keyValueStorage.go:37]: protected all the way

File: process/smartContract/builtInFunctions/saveUserName.go (hunk 7 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/saveUserName.go:46]: protected all the way

File: process/smartContract/process.go (hunk 8 above)
Comment (sasurobert) [reply process/smartContract/process.go:190]: protected all the way

File: process/transaction/transactionCostEstimator.go (hunk 9 above)
Comment (sasurobert) [reply process/transaction/transactionCostEstimator.go:49]: protected all the way

File: vm/process/systemVM.go (hunk 10 above)
Comment (sasurobert) [reply vm/process/systemVM.go:139]: protected all the way

File: vm/systemSmartContracts/auction.go (hunk 11 above)
Comment (sasurobert) [reply vm/systemSmartContracts/auction.go:1440]: protected all the way

File: vm/systemSmartContracts/esdt.go (hunk 12 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:800]: protected all the way

File: vm/systemSmartContracts/governance.go (hunk 13 above)
Comment (sasurobert) [reply vm/systemSmartContracts/governance.go:915]: protected all the way

File: vm/systemSmartContracts/staking.go (hunk 14 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1432]: protected all the way

File: process/smartContract/builtInFunctions/esdtBurn.go [hunk 21]
```diff
@@ -45,11 +47,21 @@ func NewESDTBurnFunc(
 	return e, nil
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (e *esdtBurn) SetNewGasConfig(gasCost *process.GasCost) {
+	e.mutExecution.Lock()
+	e.funcGasCost = gasCost.BuiltInCost.ESDTBurn
+	e.mutExecution.RUnlock()
```
Comment (SebastianMarian) [inline process/smartContract/builtInFunctions/esdtBurn.go:54]: Unlock() instead RUnlock()

File: process/smartContract/builtInFunctions/esdtTransfer.go [hunk 22]
```diff
@@ -51,11 +53,21 @@ func NewESDTTransferFunc(
 	return e, nil
 }
 
+// SetNewGasConfig is called whenever gas cost is changed
+func (e *esdtTransfer) SetNewGasConfig(gasCost *process.GasCost) {
+	e.mutExecution.Lock()
+	e.funcGasCost = gasCost.BuiltInCost.ESDTBurn
```
Comment (SebastianMarian) [inline process/smartContract/builtInFunctions/esdtTransfer.go:59]: ESDTBurn or ESDTTransfer?

File: process/smartContract/process.go [hunk 23]
```diff
@@ -148,7 +150,7 @@ func NewSmartContractProcessor(args ArgsNewSmartContractProcessor) (*scProcessor
 		return nil, process.ErrNilEpochNotifier
 	}
 
-	apiCosts := args.GasSchedule[core.ElrondAPICost]
+	apiCosts := args.GasSchedule.LatestGasSchedule()[core.ElrondAPICost]
```
Comment (SebastianMarian) [inline process/smartContract/process.go:153]: args.GasSchedule.LatestGasSchedule() should not be check for nil before ?

File: vm/factory/systemSCFactory.go [hunk 24]
```diff
@@ -120,6 +124,24 @@ func (scf *systemSCFactory) createGasConfig(gasMap map[string]map[string]uint64)
 	return nil
 }
 
+// GasScheduleChange is called when gas schedule is changed, thus all contracts must be updated
+func (scf *systemSCFactory) GasScheduleChange(gasSchedule map[string]map[string]uint64) {
```
Comment (SebastianMarian) [inline vm/factory/systemSCFactory.go:128]: It is not needed a mutExecution also here?

File: vm/interface.go [hunk 25]
```diff
@@ -10,6 +10,7 @@ import (
 // SystemSmartContract interface defines the function a system smart contract should have
 type SystemSmartContract interface {
 	Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
+	SetNewGasCosts(gasCosts GasCost)
```
Comment (SebastianMarian) [inline vm/interface.go:13]: SetNewGasCost(gasCost GasCost) ?

File: vm/mock/systemSCStub.go [hunk 26]
```diff
@@ -1,13 +1,21 @@
...
 // SystemSCStub -
 type SystemSCStub struct {
-	ExecuteCalled func(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
-	ValueOfCalled func(key interface{}) interface{}
+	ExecuteCalled        func(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
+	SetNewGasCostsCalled func(gasCost vm.GasCost)
+}
+
+// SetNewGasCosts -
+func (s *SystemSCStub) SetNewGasCosts(gasCost vm.GasCost) {
```
Comment (SebastianMarian) [inline vm/mock/systemSCStub.go:15]: SetNewGasCost here and in the comment

File: vm/mock/systemSCStub.go [hunk 27]
```diff
@@ -1,13 +1,21 @@
...
+	"github.com/ElrondNetwork/elrond-go/vm"
 	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
 )
 
 // SystemSCStub -
 type SystemSCStub struct {
-	ExecuteCalled func(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
-	ValueOfCalled func(key interface{}) interface{}
+	ExecuteCalled        func(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
+	SetNewGasCostsCalled func(gasCost vm.GasCost)
```
Comment (SebastianMarian) [inline vm/mock/systemSCStub.go:11]: SetNewGasCostCalled

File: config/config.go [hunk 28]
```diff
@@ -429,3 +430,14 @@ type VersionsConfig struct {
 	VersionsByEpochs []VersionByEpochs
 	Cache            CacheConfig
 }
+
+// GasScheduleByEpochs represents a gas schedule toml entry that will be applied between from the provided epoch
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply config/config.go:434]: done

File: core/forking/gasSchedule.go [hunk 29]
```diff
@@ -0,0 +1,134 @@
...
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	sort.Slice(g.gasScheduleConfig.GasScheduleByEpochs, func(i, j int) bool {
+		return g.gasScheduleConfig.GasScheduleByEpochs[i].StartEpoch < g.gasScheduleConfig.GasScheduleByEpochs[j].StartEpoch
+	})
+	var err error
+	g.lastGasSchedule, err = core.LoadGasScheduleConfig(filepath.Join(configDir, gasScheduleConfig.GasScheduleByEpochs[0].FileName))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/forking/gasSchedule.go:48]: no. it has to take the first one. epoch notifier will give later the correct epoch and will set to the needed value.

File: core/forking/gasSchedule.go [hunk 30]
```diff
@@ -0,0 +1,134 @@
...
+	handlers          []core.GasScheduleSubscribeHandler
+}
+
+// NewGasScheduleNotifier creates a new instance of a gasScheduleNotifier component
+func NewGasScheduleNotifier(
+	gasScheduleConfig config.GasScheduleConfig,
+	configDir string,
+	epochNotifier core.EpochNotifier,
+) (*gasScheduleNotifier, error) {
+	g := &gasScheduleNotifier{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/forking/gasSchedule.go:32]: done

**Review by sasurobert — COMMENTED** [review]
common github

File: vm/mock/systemSCStub.go (hunk 27 above)
Comment (sasurobert) [reply vm/mock/systemSCStub.go:11]: done

File: vm/mock/systemSCStub.go (hunk 26 above)
Comment (sasurobert) [reply vm/mock/systemSCStub.go:15]: done

File: vm/interface.go (hunk 25 above)
Comment (sasurobert) [reply vm/interface.go:13]: done

File: vm/factory/systemSCFactory.go (hunk 24 above)
Comment (sasurobert) [reply vm/factory/systemSCFactory.go:128]: no - this is a factory.

File: process/smartContract/process.go (hunk 23 above)
Comment (sasurobert) [reply process/smartContract/process.go:153]: done

File: process/smartContract/builtInFunctions/esdtTransfer.go (hunk 22 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtTransfer.go:59]: done

File: process/smartContract/builtInFunctions/esdtBurn.go (hunk 21 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtBurn.go:54]: done

File: genesis/process/genesisBlockCreator.go [hunk 31]
```diff
@@ -171,7 +171,7 @@ func checkArgumentsForBlockCreator(arg ArgsGenesisBlockCreator) error {
 	if check.IfNil(arg.AccountsParser) {
 		return genesis.ErrNilAccountsParser
 	}
-	if arg.GasMap == nil {
+	if arg.GasSchedule == nil {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply genesis/process/genesisBlockCreator.go:174]: done

File: core/forking/gasSchedule.go [hunk 32]
```diff
@@ -51,13 +64,13 @@ func NewGasScheduleNotifier(
 
 // RegisterNotifyHandler will register the provided handler to be called whenever a new epoch has changed
 func (g *gasScheduleNotifier) RegisterNotifyHandler(handler core.GasScheduleSubscribeHandler) {
-	if check.IfNil(handler) {
+	if check.IfNilReflect(handler) {
```
Comment (iulianpascalau) [inline core/forking/gasSchedule.go:67]: why this change?

File: vm/systemSmartContracts/auction.go [hunk 33]
```diff
@@ -44,6 +45,8 @@ type stakingAuctionSC struct {
 	marshalizer        marshal.Marshalizer
 	flagStake          atomic.Flag
 	flagAuction        atomic.Flag
+
+	mutExecution sync.Mutex
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/auction.go:49]: Please indent this

File: vm/systemSmartContracts/auction.go [hunk 34]
```diff
@@ -132,6 +135,9 @@ func NewStakingAuctionSmartContract(
 
 // Execute calls one of the functions from the auction staking smart contract and runs the code according to the input
 func (s *stakingAuctionSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	s.mutExecution.Lock()
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/auction.go:138]: RLock?

Comment (sasurobert) [reply vm/systemSmartContracts/auction.go:138]: done

File: vm/systemSmartContracts/auction.go (hunk 33 above)
Comment (sasurobert) [reply vm/systemSmartContracts/auction.go:49]: done

File: core/forking/gasSchedule.go (hunk 32 above)
Comment (sasurobert) [reply core/forking/gasSchedule.go:67]: it does not have isInterfaceNil. as that method is not pushed to vmcommon.

File: core/errors.go [hunk 35]
```diff
@@ -82,3 +82,6 @@ var ErrFileLoggingProcessIsClosed = errors.New("file logging process is closed")
 
 // ErrNilFeeConfig signals that a nil fee config is provided
 var ErrNilFeeConfig = errors.New("nil fee config")
+
+// ErrInvalidGasScheduleConfigs signals that invalid gas schedule config was provided
+var ErrInvalidGasScheduleConfigs = errors.New("invalid gas schedule config")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/errors.go:87]: done

File: process/smartContract/process.go [hunk 36]
```diff
@@ -134,7 +134,7 @@ func NewSmartContractProcessor(args ArgsNewSmartContractProcessor) (*scProcessor
 	if check.IfNil(args.GasHandler) {
 		return nil, process.ErrNilGasHandler
 	}
-	if args.GasSchedule == nil {
+	if args.GasSchedule == nil || args.GasSchedule.LatestGasSchedule() == nil {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:137]: done

**Review by SebastianMarian — APPROVED** [review]

File: core/forking/gasSchedule.go (hunk 32 above)
Comment (iulianpascalau) [reply core/forking/gasSchedule.go:67]: ok

**Review by iulianpascalau — APPROVED** [review]
